File: citrussoda/8week/208_implement-trie.cpp
class TrieNode {
 public:
  // 알파벳 가르킬 포인터
  TrieNode *child[26];
  // 해당 노드까지 완성된 단어인가 확인
  bool isWord;

  TrieNode() {
    isWord = false;
    for (auto &a : child) a = nullptr;
  }
};

class Trie {
  TrieNode *root;

 public:
  Trie() { root = new TrieNode(); }

  void insert(string s) {
    TrieNode *p = root;

    for (auto &a : s) {
      int i = a - 'a';
      if (!p->child[i]) p->child[i] = new TrieNode();
      p = p->child[i];
    }
    // 다 할당했으면 단어라고 선언, 포인터 변수기 때문에 for문에서 끝까지 할당한
    // 노드에서 isWord가 true
    p->isWord = true;
  }

  bool search(string key, bool prefix = false) {
    TrieNode *p = root;

    for (auto &a : key) {
      int i = a - 'a';
      if (!p->child[i]) return false;
      p = p->child[i];
    }

    // startWith를 위한 prefix 추가, startWith가 아니면 해당 노드가 끝까지
    // 갔는지 isWord만 확인하면 된다.
    if (prefix == false) return p->isWord;

    // startWith라면 해당 노드까지 모두 있는 것이므로 true
    return true;
  }

  bool startsWith(string prefix) { return search(prefix, true); }
};

File: wjsdncl/208_implement_trie.js
/**
 * 24m12s 소요
 *
 * 시간복잡도: O(m) (m: 문자열의 길이)
 *
 * Trie 구조를 처음 접해봐서 풀이와 ChatGPT의 도움을 받았습니다.
 *
 * Trie 구조를 구현하는 문제입니다.
 * Trie 구조는 문자열을 저장하고 검색하는 데 사용되는 트리 자료구조입니다.
 */
var TrieNode = function () {
  this.children = {};
  this.isEnd = false;
};

var Trie = function () {
  this.root = new TrieNode();
};

Trie.prototype.insert = function (word) {
  let node = this.root;
  for (let char of word) {
    if (!node.children[char]) {
      node.children[char] = new TrieNode();
    }
    node = node.children[char];
  }
  node.isEnd = true;
};

Trie.prototype.search = function (word) {
  let node = this.root;
  for (let char of word) {
    if (!node.children[char]) {
      return false;
    }
    node = node.children[char];
  }
  return node.isEnd;
};

Trie.prototype.startsWith = function (prefix) {
  let node = this.root;
  for (let char of prefix) {
    if (!node.children[char]) {
      return false;
    }
    node = node.children[char];
  }
  return true;
};
